## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [7]:
arr = np.random.randint(0, 20, size=10)
arr[(arr > 3) & (arr < 8)] *= -1
arr

array([15, -4, 13, 15, 17, 13,  3, 13,  0, -6])

**2.** Заменить максимальный элемент случайного массива на 0

In [8]:
arr = np.random.randint(0, 20, size=10)
arr[np.argmax(arr)] = 0
arr

array([ 8,  5, 11,  5,  0, 18,  1,  6, 12,  7])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [9]:
arrays = np.array([[1, 2], [3, 4], [5, 6]])

result = np.array(np.meshgrid(*arrays)).T.reshape(-1, len(arrays))
result


array([[1, 3, 5],
       [1, 4, 5],
       [2, 3, 5],
       [2, 4, 5],
       [1, 3, 6],
       [1, 4, 6],
       [2, 3, 6],
       [2, 4, 6]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [22]:
A = np.array([[4, 1, 3],
              [7, 8, 1],
              [3, 2, 0],
              [4, 3, 1],
              [2, 3, 1],
              [6, 4, 3],
              [3, 1, 2],
              [9, 5, 4]])

B = np.array([[1, 2], [3, 4]])

def contains_all_elements(row, b_row):
    return all(elem in row for elem in b_row)

matching_rows = []
for i, row in enumerate(A):
    if all(any(contains_all_elements(row, b_row) for b_row in B) for b_row in B):
        matching_rows.append(i)


print(matching_rows)


[0, 3, 4, 5, 6]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [44]:
arr = np.array([[4, 1, 3],
                   [1, 1, 1],
                   [3, 2, 0],
                   [4, 3, 1]])
dic  = {}
for idx, row in enumerate(arr):
    dic[idx] = row.tolist()


result = []

for key, value in dic.items():
    if value and len(set(value)) > 1:
        result.append(value) 

result_array = np.array(result)

print(dic)

{0: [4, 1, 3], 1: [1, 1, 1], 2: [3, 2, 0], 3: [4, 3, 1]}


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [48]:
arr = np.array([[4, 1, 3],
                   [1, 1, 1],
                   [1, 1, 1],
                   [4, 3, 1]])
dic  = {}
for idx, row in enumerate(arr):
    dic[idx] = row.tolist()


result = []

for key, value in dic.items():
        if value not in result:
            result.append(value) 

result_array = np.array(result)

result_array


array([[4, 1, 3],
       [1, 1, 1],
       [4, 3, 1]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [54]:
#1вариант
X = np.array([[1, 0, 1], 
              [2, 0, 2], 
              [3, 0, 3],
               [4,4,4] ])

result = np.prod(np.diag(X)[np.diag(X)!=0]  )
result

#2
X = [[1, 0, 1], 
    [2, 0, 2], 
    [3, 0, 3],
    [4, 4, 4] ]
product = 1
found_non_zero = False  

rows = len(X)
cols = len(X[0])


for i in range(min(rows, cols)):
    if X[i][i] != 0:  
        product *= X[i][i]  
        found_non_zero = True

if not found_non_zero:
    product = 0

print(product)





Произведение ненулевых элементов на диагонали: 3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [56]:

x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

unique_x, counts_x = np.unique(x, return_counts=True)
unique_y, counts_y = np.unique(y, return_counts=True)

are_equal = np.array_equal(unique_x, unique_y) and np.array_equal(counts_x, counts_y)

print( are_equal)

Задают ли векторы одно и то же мультимножество: True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [60]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

zero = x==0
print (x[1:][zero[:-1]].max())

5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [62]:

arr = np.array([1, 2, 2, 3, 4, 2, 5])
unique_elements, counts = np.unique(arr, return_counts=True)



array([1, 3, 1, 1, 1])

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [66]:
import time

def euclidean_distance(X, Y):
    return np.sqrt(np.sum((X[:, np.newaxis, :] - Y[np.newaxis, :, :]) ** 2, axis=-1))

X = np.random.rand(100, 5)  
Y = np.random.rand(80, 5)   

start_time = time.time()
distances_np = euclidean_distance(X, Y)
np_time = time.time() - start_time

distances_np


array([[1.08473649, 1.12441745, 1.1444759 , ..., 1.07809765, 1.04147479,
        0.58848188],
       [0.94129786, 0.83218264, 0.81100109, ..., 0.70204721, 0.75905901,
        0.57046822],
       [1.1472336 , 1.043064  , 0.85193251, ..., 0.8333967 , 0.96069878,
        0.69998208],
       ...,
       [0.78163743, 0.80132627, 0.86941861, ..., 0.81562375, 0.73666688,
        0.53597011],
       [0.99786806, 0.87280144, 0.85029936, ..., 0.90935988, 0.66620604,
        1.00035407],
       [0.8105148 , 0.96778613, 1.09760648, ..., 0.97282774, 0.93037037,
        0.59133751]])

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [67]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [68]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [70]:
average_calories = calorie_stats.mean()
average_calories

np.float64(106.88311688311688)

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [74]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [75]:
median_calories = np.median(calorie_stats)
median_calories

np.float64(110.0)

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [76]:

nth_percentile = None

for p in range(101):
    percentile_value = np.percentile(calorie_stats, p)
    if percentile_value > 60:
        nth_percentile = percentile_value
        print(f"Наименьший процентиль, превышающий 60 калорий: {nth_percentile} (процентиль {p})")
        break

Наименьший процентиль, превышающий 60 калорий: 70.0 (процентиль 4)


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [80]:
more_calories = (np.size(calorie_stats[calorie_stats > 60])/np.size(calorie_stats))*100
more_calories

96.1038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [81]:
calorie_std = np.std(calorie_stats)
calorie_std

np.float64(19.35718533390827)

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Полученные данные говорят о следующем:

Большинство конкурентов (96% +) имеют больше калорий, чем CrunchieMunchies
Большое стандратное отклонение показывает то, что данные в выборке достаточно сильно разнятся, но, учитывая вышеописанный пункт, этот факт наоборот приобретает положительный окрас, т.к. отклонение в сторону, которая больше среднего по выборке, является недостатком конкурентов.
Разница калорийности CrunchieMunchies и медианы по выборке составляет 83% от калорийности CrunchieMunchies, что тоже можно упомянуть в виде преимущества.
В целом CrunchieMunchies классные хлопья, но лучше конечно пиво.